### PR TITLE
Python3 signed

### PR DESCRIPTION
**Pull request by codeix**
Hi everyone,

This is a new pull request replacing the old one https://github.com/zopefoundation/martian/pull/1 . In contrast to the previous request, this one contains clean and signed commit messages.
If nobody against it, i will merge this request to master soon.

Kind regards,
codeix

### CODE REVIEW

**Comment by mgedmin** [comment]
Curious: what do you mean when you say "signed commit messages"?

**Comment by codeix** [comment]
In the first pull request I was using different e-mail addresses, I now streamlined this to one. In addition, every commit has a pgp signature as per http://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work  but I am not sure whether this really was required.

**Comment by mgedmin** [comment]
Oh, nice.  No, we don't require GPG signatures.

**Comment by mgedmin** [comment]
(BTW you can force-push the same branch after rebasing to update the old pull request, it's not necessary to open a new one every time.)

File: src/martian/directive.py [hunk 1]
```diff
@@ -147,7 +147,10 @@ def _default(mro, get_default):
     # if we haven't found a result, raise the first error we had as
     # a GrokError
     if error is not None:
-        raise GrokError(unicode(error), error.component)
+        if sys.version_info[0] < 3:
+            raise GrokError(unicode(error), error.component)
+        else:
+            raise GrokError(str(error), error.component)
```
Comment (mgedmin) [inline src/martian/directive.py]: I'd rather avoid an if here and instead have something like [six.text_type](https://pythonhosted.org/six/#six.text_type) that' an alias for unicode on Python 2 and str on Python 3.  You could define it in compat3.py, if you don't want to introduce a dependency on six.

File: CHANGES.txt [hunk 2]
```diff
@@ -4,8 +4,13 @@ CHANGES
 0.15 (unreleased)
 =================
 
+- compatibility for python 3
+- update code to work with python2 and python3
+- update doctests to work with python2 and python3
+- adjust egg to work with newer version of setuptools
```
Comment (mgedmin) [inline CHANGES.txt]: I think "compatibility for Python 3" implies that code and doctests were updated to work with it.

File: setup.py [hunk 3]
```diff
@@ -17,7 +17,7 @@ def read(*rnames):
 
 setup(
     name='martian',
-    version='0.15dev',
+    version = '0.15.dev0',
```
Comment (mgedmin) [inline setup.py]: PEP-8 says no spaces around `=`.

File: setup.py [hunk 4]
```diff
@@ -28,11 +28,20 @@ def read(*rnames):
 (http://grok.zope.org)
 """,
     long_description=long_description,
+    classifiers= [
```
Comment (mgedmin) [inline setup.py]: PEP-8 says no space after the `=`.

File: src/martian/directive.py [hunk 5]
```diff
@@ -246,8 +249,8 @@ def check_factory_signature(self, *arguments, **kw):
         argspec = inspect.formatargspec(args[1:], varargs, varkw, defaults)
         exec("def signature_checker" + argspec + ": pass")
         try:
-            signature_checker(*arguments, **kw)
-        except TypeError, e:
+            locals()['signature_checker'](*arguments, **kw)
```
Comment (mgedmin) [inline src/martian/directive.py]: This bit confuses me.

File: src/martian/scan.txt [hunk 6]
```diff
@@ -20,7 +20,7 @@ information about a module or a package. Let's take a look at the
 
   >>> module_info = module_info_from_dotted_name('martian.tests.stoneage.cave')
 
-We get a ``ModuleInfo`` object representing the ``cave module::
+We get a ``ModuleInfo`` object representing the ``cave module``::
```
Comment (mgedmin) [inline src/martian/scan.txt]: Good catch!

File: src/martian/testing.py [hunk 7]
```diff
@@ -62,5 +62,15 @@ def __init__(cls, classname, bases, dict_):
...
+    class FakeModuleObject(object):
+        pass
+    
+    class FakeModule(object):
+        __metaclass__ = FakeModuleMetaclass
+else:
+    from martian.testing_compat3 import FakeModule
+    from martian.testing_compat3 import FakeModuleObject
+
+
```
Comment (mgedmin) [inline src/martian/testing.py]: [six.with_metaclass](https://pythonhosted.org/six/#six.with_metaclass) would come in handy here.

File: src/martian/util.py [hunk 8]
```diff
@@ -22,21 +22,30 @@
...
 
 import martian
+from martian.compat3 import CLASS_TYPES
 from martian.error import GrokError, GrokImportError
 
 def not_unicode_or_ascii(value):
-    if isinstance(value, unicode):
+
+    # python3 compatibility
+    if sys.version_info < (3,) and  isinstance(value, unicode):
```
Comment (mgedmin) [inline src/martian/util.py]: There are two spaces after `and`.

I wonder if a `text_type` alias would make this simpler?

**Comment by mgedmin** [comment]
LGTM.  All the comments are minor.

**Comment by mgedmin** [comment]
Once this is merged we'll want a new Martian release on PyPI.  The only people who can upload a release are:

> Package Index Owner: faassen, jw, philikon

so: pinging @faassen, @janwijbrand, @philikon.

(If you want me to make a release, please grant PyPI rights to `mgedmin`.)

File: src/martian/directive.py (hunk 5 above)
Comment (codeix) [inline src/martian/directive.py]: This confuses me too! 
Without the locals() i get this error in python 3: NameError: name 'signature_checker' is not defined
With my last commit the code is using a namespace, so its a little bit nicer...
I wonder if possible to work without dynamic code?

**Comment by faassen** [comment]
@mgedmin, I've just given you the owner role for Martian.

File: src/martian/testing.py (hunk 7 above)
Comment (codeix) [inline src/martian/testing.py]: The __qualname__ seem not be handled by six.with_metaclass

**Comment by mgedmin** [comment]
LGTM.

I don't want to merge now because I'm too tired.  Please ping me if I forget to get back to this in a reasonable timeframe.  Or just merge yourself, since you're now a committer ;-).

**Comment by codeix** [comment]
Just merged the pull request. @mgedmin thanks for all your help!

**Comment by mgedmin** [comment]
Version 0.15 released to PyPI.

**Pull request merged**